File: Leetcode/0077. Combinations/0077.cpp
class Solution {
 public:
  vector<vector<int>> combine(int n, int k) {
    vector<vector<int>> ans;
    dfs(n, k, 1, {}, ans);
    return ans;
  }

 private:
  void dfs(int n, int k, int s, vector<int>&& path, vector<vector<int>>& ans) {
    if (k == 0) {
      ans.push_back(path);
      return;
    }

    for (int i = s; i <= n; ++i) {
      path.push_back(i);
      dfs(n, k - 1, i + 1, move(path), ans);
      path.pop_back();
    }
  }
};

File: Leetcode/0077. Combinations/0077.py
class Solution:
  def combine(self, n: int, k: int) -> List[List[int]]:
    ans = []

    def dfs(n: int, k: int, s: int, path: List[int]) -> None:
      if k == 0:
        ans.append(path.copy())
        return

      for i in range(s, n + 1):
        path.append(i)
        dfs(n, k - 1, i + 1, path)
        path.pop()

    dfs(n, k, 1, [])
    return ans
